### ISSUE TITLE
Blank screen + page keeps refreshing? 

### ISSUE THREAD

**Issue by michaelsoriano** [devloco/create-react-wptheme#17]
Hi @devloco - a couple of observations. There would be times when I'm in wpstart mode, I save a file, browser refreshes and its just a blank screen. I've also seen the page keeps refreshing multiple times - non stop. I'm guessing it might be chrome. 

It's not a big deal, all I need to do is npm run wpstart again - and it seems to fix itself. 

Also an update: I'm almost done with my project and I must tell you that I'm really liking this setup. It seems that this is better than having separate servers for the front and back - because this way, we can keep some of the WP functionality and the React app - in the same url. 

For example, I have my a couple of WP pages that are using theme (template files) and I have the main React app. Its pretty sweet! Thanks for all your work on this.

**Comment by devloco** [comment]
Hiya @michaelsoriano, I've seen the blank screen thing myself. Usually just re-saving a file, forcing it to rebuild, should be enough... I haven't needed to stop/restart it before. I think it happens when you save a file while it's building the previous save.

I've never seen the constant reloading, though maybe a rebuild or just a hard refresh would stop it...?(Chrome: SHIFT+F5).

As always, thanks for the kind words. I must admit I never understood the whole "separation of concerns" over multiple servers thing. I mean if your front-end server is down, then what good is it that the backend is working? And vice-versa.

Anyway, I'm happy to hear it worked out for ya. If you decide to write-up a tutorial and post it somewhere, I'll link to it from the README. (hint hint ;-))

**Comment by michaelsoriano** [comment]
@devloco  - You can discard the "Page keeps refreshing" issue. It turns out that I had 2 tabs open for the site. This causes it to refresh over and over. So I just kill one of the tabs and its fine. 

Yes - I am writing a tutorial and will let you know when its up.

**Comment by devloco** [comment]
I think I've fixed the constant refresh when multiple tabs are open. I actually consider that a feature (although I discovered it by accident). What's cool about it is that you can have multiple browsers open while developing your theme (say Chrome and Firefox) and both browsers will refresh, thus allowing you to see if the theme renders the same in both browsers. Sort of a makeshift cross-browser testing tool.

I'm going to go ahead and close this ticket, but if someone can figure out a reproducible set of steps for the blank screen issue, I'd love to figure that out too. But since I can't make it happen on command, it's hard to know what to fix.

Either of these simple workarounds should work if you encounter it:

1. Just manually refresh the browser. Chances are the currently built files are in place, the browser just got the previous build.
2. If that doesn't work, make a simple change to a file that will force Webpack to rebuild the theme again. If your browser doesn't auto-refresh in that case, you'll just have to manually refresh it once.

Thanks for the bug-hunting!

**Comment by michaelsoriano** [comment]
That's fine @devloco, thanks for looking into it. 

But maybe this is worth a quick look - because it just happened to me just now.

For some reason, as soon as react-router is installed and added the code below, it kept happening a lot:

`npm install react-router-dom`

in index.js, include: 

`import { Route, BrowserRouter as Router, Switch } from 'react-router-dom'`

then in index.js:

```
const routes = (
    <Router>
      <Switch>
          <Route path="/" component={App} />         
      </Switch>
    </Router>
  )

ReactDOM.render(routes, document.getElementById('root'));
```

I try to add other routes to other components and still a blank screen. 

I also do ctrl + F5, many times - but no luck.

**Comment by devloco** [comment]
Oh, that's great actually. If I can reproduce it, I can fix it! Thanks!

**Comment by devloco** [comment]
Hi @michaelsoriano,

I just tried this real quick, and it's not reproducing for me... :(

A couple of things come to mind:

- Is this an existing theme that's been updated?
- Or maybe you've made changes to the `user.dev.json` file?

I recently (been a couple of months actually) moved the injection location of the wptheme `<script>` tags. Are yours getting injected in a different spot in `index.php` than shown in this screenshot:

[View Source from Firefox](https://github.com/devloco/create-react-wptheme/blob/master/screenshots/view-source.png)

Basically if those tags are after the scripts that Webpack inject (e.g. `bundle.js`, etc), then they can't do their job.

Otherwise, I'll try some more on my end as well, later.

Thanks.

**Comment by devloco** [comment]
Hi @michaelsoriano,

I'm trying some other things now, and I remembered from your tutorial that you use Git Bash. I decided to try that too. It's probably not the cause of the blank screen problem, but all of Webpack's output to the command prompt is missing on each build.

Here's what it looks like in a standard Windows command prompt. I know it's a reach, but maybe there's an error happening, and it's not being displayed in Git Bash...?

![2019-05-29 16_11_10](https://user-images.githubusercontent.com/40293912/58591672-6bc8bb80-822c-11e9-83f4-8ad37e60c7f1.png)

**Comment by devloco** [comment]
You probably leave your JS console open while developing, but just in case you don't... you might check there too. If there's a runtime error, it might slip past the error overlay. But nothing gets past the JS console.

**Comment by michaelsoriano** [comment]
Hi @devloco, 

So here are more details: 

- Is this an existing theme that's been updated? - Its the same repo ["barebones"](https://github.com/michaelsoriano/barebones) that I started writing a tutorial on. I just committed all changes - in case you want to take a look at the code. 
- Or maybe you've made changes to the user.dev.json file? no, I haven't updated that.
- The location of the wpThemeClient.start is the same as your screenshot
- Yes, I'm using Git Bash - but seems that no error is displaying. Not sure. 
- No errors in the JS console. 

I created a small animated gif to demonstrate. You will see that it is is sporadic - it goes blank, and sometimes it shows up, but changes are not reflected. 

![blank-screen](https://user-images.githubusercontent.com/5881081/58647546-abd77f00-82bc-11e9-8610-4594111341c9.gif)

Notice when I change the source (I added the text "some other change") - it doesn't reflect it in the browser - but this may be another issue. 

Then it goes blank again - even with repeated manual refresh.  the terminal shows "Browser refresh sent" and the console is "The browser refresh server is connected."

I'm on a Windows 10, using Wamp Server and Chrome - if that matters. 

Thanks for looking into it.

**Comment by devloco** [comment]
Well, this really sucks.
I changed all my code to look like yours, and even switched to Chrome. But I haven't seen a single blank screen yet....

I hate to waste your time with a bunch of "try this" or "try that" messages... but if you're willing...

**TL;DR**: your gif sure makes it look like a caching issue. Maybe from the browser... or maybe some server caching tool built into Wamp...? FYI: That's the only difference, that I'm aware of, in our setups now, I'm using Xampp.

**So what I'm going to do** is code in a "cache buster" ... basically the JS files will have different names every time you save. Then Wamp and Chrome will be forced to load them every time.

Of course, since I can't reproduce the issue on my end... there's no guarantee it'll fix up your stuff. :-(
But you can try to disable caching in Chrome...

But if Wamp is doing any caching... I don't know how to check or disable that.

Here's where you can turn off all caching in Chrome while developing:

![2019-05-31 12_34_48](https://user-images.githubusercontent.com/40293912/58724946-82455300-83a3-11e9-944b-2fc5721b63c0.png)

**Here's the long winded version of what I see in your animated gif.**

- The first frame of the gif shows a blank browser.
    - Remember, carriage returns and spaces do not matter. All that matters are the actual characters in the code.
    - Every time your browser shows as blank in the gif, your code is considered to be the exact same as it is on frame 1 of the gif.
    - Again, carriage returns and any other types of "white space" (spaces, tabs, etc) do not matter to Webpack when it builds the files.
- **And then I see some holes in my theory**. Because, as you said, when your code is actually different, the changes aren't actually appearing! Argh...
- I've tried everything I can think of to try and match what you're seeing, but everything I try just works correctly.

**Comment by devloco** [comment]
PS: I'll work on the cache busting code over the weekend. But may not get it done until next week.

**Comment by devloco** [comment]
PPS: just to be clear, the "this is bad" comment in the screen shot above refers to both red boxes, not just the one the arrow is pointing at.

**Comment by adigitalnative** [comment]
Since the two of us are seeing this behavior on occasion (I’ve been fixing it by saving a file and letting the browser reload - most of the time this seems to take care of it. It’s not regular enough that I can easily reproduce)

Michael, what is your stack within your theme?

Mine:

react
react-dom
react-redux
react-render-html
react-router-dom
redux
redux-thunk
semantic-ui
semantic-ui-css
semantic-ui-react

I’m wondering if this might be some kind of Redux issue? I believe I saw it more frequently when I was fiddling with my actions in redux…

---------------------------

Jacqueline Chenault
jq@nativefoundry.com

202.489.6128
Github | Linkedin | Portfolio

---------------------------

**Comment by devloco** [comment]
Thanks for jumping in @adigitalnative. Your description sounds more like what I would expect when this happens. Basically the browser got served an "old build" of the files. And just forcing Webpack to rebuild and reload fixes it.

Unfortunately what I'm understanding from @michaelsoriano, it's not the same. Just forcing a rebuild doesn't seem to help for him.

**Comment by devloco** [comment]
I'm certainly not going to dismiss anything at this point, but I'll be surprised if it is caused by any particular package. But I would guess that the number of packages, and the combined "weight" of those packages could have an effect. Especially from @adigitalnative 's description.

And I'll stand by my initial hunch after watching @michaelsoriano animated gif a hundred times. It looks like a caching issue for him.

It could be that a blank page is a symptom of more than 1 problem.

**Comment by michaelsoriano** [comment]
@devloco - That "disable cache" checkbox is definitely working so far. I haven't seen a white screen since. 
I'm going to continue building the theme and will monitor the white screen. Hopefully that's it. 
(I feel stupid not knowing about that checkbox. I thought doing a CTRL + F5 was enough)

@adigitalnative - I have the following modules:

- react
- react-dom
- react-router-dom

Thanks for really digging into this. Your explanations above really helped out. That's some good debugging skills (considering you couldn't even replicate it).

**Comment by devloco** [comment]
@michaelsoriano that is great news! I was feeling pretty pessimistic about this issue. But if the disable cache checkbox works for you, then a "cache buster" should work for any new users that come along and have caching enabled. It makes for a better first impression for sure.

Speaking of cache busting, I just published that change to NPM.

It's a pretty major change to the code, so I'll be surprised if I didn't create a new bug or two. But so far it's working in my testing. It does fill up the "build" and "static/js" folders. But all of the "extra" files are wiped out each time you run "npm run wpstart" or "npm run wpbuild".

Obligatory link to the updating instructions, [here](https://github.com/devloco/create-react-wptheme#updating-existing-themes).

@adigitalnative , please give the new version a try... I'm hoping it works for your situation as well.

Thanks to you both! Have a good weekend.

**Comment by michaelsoriano** [comment]
@devloco  i updated to the latest and looks like a good fix! I don't even have to check the "disable cache" and its reloading the changes everytime - no blank screen! 

 ![blank-page-fix](https://user-images.githubusercontent.com/5881081/58751039-3d2a2b00-844e-11e9-81cc-3acae46b5c1a.gif)

That cache buster is definitely a good feature. I'm going to close this issue. Thanks again!

**Comment by devloco** [comment]
@michaelsoriano, I’m out of town this week, but I wanted jump on and thank you for your patience and help with this one. Glad we got it figured out. 

Seeing it get closed was a huge relief.

**Comment by CalvinJamesHeath** [comment]
Hi, I'm experiencing the same issue; The react app doesn't stop refreshing itself on chrome. Even after I delete everything that app.js is rendering.

**Comment by liamosycka** [comment]
@CalvinJamesHeath Had the same error, I solved it using firefox extended support release... I still cannot open it in chrome.
